theano/tensor/nnet/nnet.py

```diff
@@ -130,329 +132,496 @@ def c_code_template(dtype):
     __props__ = ()
 
-    def make_node(self, x):
+    def make_node(self, x, axis=-1):
         x = tensor.as_tensor_variable(x)
-        if x.type.ndim not in (1, 2) \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
-                             x.type)
+        axis = tensor.as_tensor_variable(axis)
+        # TODO : Delete this and modify the test accordly
         if x.ndim == 1:
             x = tensor.shape_padleft(x, n_ones=1)
-
-        return Apply(self, [x], [x.type()])
+        inputs = [x, axis]
+        return Apply(self, inputs, [x.type()])
 
     def perform(self, node, input_storage, output_storage):
-        x, = input_storage
-        e_x = np.exp(x - x.max(axis=1)[:, None])
-        sm = e_x / e_x.sum(axis=1)[:, None]
+        x, axes = input_storage
+        # Apply softmax on the specified dimension
+        e_x = np.exp(x - x.max(axis=axes, keepdims=True))
+        sm = e_x / e_x.sum(axis=axes, keepdims=True)
         output_storage[0][0] = sm
 
     def L_op(self, inp, outputs, grads):
-        x, = inp
-        g_sm, = grads
-        return [softmax_grad(g_sm, outputs[0])]
+        x, axes = inp
+        # To investigate
+        if len(grads) == 2:
+            g_sm, tmp = grads
+        else:
+            g_sm, = grads
+        axes_grad = theano.gradient.grad_undefined(self, 1, axes)
+        return [abstract_softmax_grad(g_sm, outputs[0], axes), axes_grad]
 
     def R_op(self, inputs, eval_points):
         # I think the Jacobian is symmetric so the R_op
         # is the same as the grad
         if None in eval_points:
-            return [None]
+            return [None, None]
+        print(eval_points)
         return self.L_op(inputs, [self(*inputs)], eval_points)
 
     def infer_shape(self, node, shape):
-        return shape
+        return [shape[0]]
+
+abstract_softmax_op = Abstract_softmax()
+
+
+class Softmax(Abstract_softmax):
+    """
+    Softmax activation function
+    :math:`\\varphi(\\mathbf{x})_j =
+    \\frac{e^{\mathbf{x}_j}}{\sum_{k=1}^K e^{\mathbf{x}_k}}`
+    where :math:`K` is the total number of neurons in the layer. This
+    activation function gets applied on the selected axis (Default axis is the last).
+    This function get applied when axis is a scalar and uses a fast C optimization.
+    """
+
+    nin = 2
+    nout = 1
+    __props__ = ()
+
+    def make_node(self, x, axis=-1):
+        x = tensor.as_tensor_variable(x)
+        axis = tensor.as_tensor_variable(axis)
+        if x.ndim == 1:
+            x = tensor.shape_padleft(x, n_ones=1)
+        # Check if axis is an integer
+        if axis.type.dtype not in tensor.int_dtypes:
+            raise ValueError('axis must be an integer. Got ', axis.type)
+        return Apply(self, [x, axis], [x.type()])
+
+    def L_op(self, inp, outputs, grads):
+        x, axes = inp
+        # To investigate
+        if len(grads) == 2:
```

Comment:
I spoke with @lamblin and as consequences of having axis as a node, we get the grad of the axis in eval_points, since this grad isn't defined we don't need to pass it through the L_op. So I put as argument eval_point[0] and it resolves the issue.
Yes, we keep the axis as input node. It simplifies the access of the axis value during the optimization. I think @lamblin can give other reasons about this choice.